tools/p4mlir-translate/translate.cpp

```diff
@@ -189,9 +190,99 @@ class P4HIRConverter : public P4::Inspector, public P4::ResolutionContext {
+        }
+        auto validityConstant = emitValidityConstant(loc, validityConstValue);
+        return builder.create<P4HIR::CmpOp>(loc, P4HIR::CmpOpKind::Eq, validityBitValue,
+                                            validityConstant);
+    }
+
+    P4HIR::CmpOp emitHeaderUnionIsValidCmpOp(mlir::Location loc, mlir::Value headerUnion,
+                                             P4HIR::HeaderUnionType headerUnionType,
+                                             P4HIR::ValidityBit validityBitValue, bool useRefOp) {
+        // Helper function to build the nested ternary operations recursively
+        std::function<mlir::Value(size_t)> buildNestedTernaryOp =
+            [&](size_t fieldIndex) -> mlir::Value {
+            // If all the fields were checked, return false
+            if (fieldIndex >= headerUnionType.getFields().size()) {
+                return getBoolConstant(loc, false);
+            }
+
+            auto fieldInfo = headerUnionType.getFields()[fieldIndex];
+            mlir::Value header;
+            if (useRefOp) {
+                header =
+                    builder.create<P4HIR::StructExtractRefOp>(loc, headerUnion, fieldInfo.name);
+            } else {
+                header = builder.create<P4HIR::StructExtractOp>(loc, headerUnion, fieldInfo.name);
+            }
+
+            // Check if this member header is valid
+            auto headerIsValid =
+                emitHeaderIsValidCmpOp(loc, header, P4HIR::ValidityBit::Valid, useRefOp);
+
+            // Create a ternary operation:
+            // if this header is valid, return true,
+            // otherwise check the next header in the header union
+            auto ternaryOp = builder.create<P4HIR::TernaryOp>(
+                loc, headerIsValid.getResult(),
+                [&](mlir::OpBuilder &b, mlir::Location loc) {
+                    // If this header is valid, return true
+                    b.create<P4HIR::YieldOp>(loc, getBoolConstant(loc, true));
+                },
+                [&](mlir::OpBuilder &b, mlir::Location loc) {
+                    // If this header is not valid, check the next header
+                    b.create<P4HIR::YieldOp>(loc, buildNestedTernaryOp(fieldIndex + 1));
+                });
+            return ternaryOp.getResult();
+        };
+
+        // Start the recursive building from the first field
+        auto isValid = buildNestedTernaryOp(0);
+
+        // Return a comparison operation for consistency with other validity checks
+        return builder.create<P4HIR::CmpOp>(loc, P4HIR::CmpOpKind::Eq, isValid,
+                                            getBoolConstant(loc, true));
+    }
+
+    void emitAllHeadersInvalidForParentHeaderUnion(mlir::Location loc, const P4::IR::Expression *header) {
+        if (const auto *member = header->to<P4::IR::Member>()) {
+            if (member->expr != nullptr && member->expr->type != nullptr) {
```

Comment:
This should never happen. In which case you're seeing `IR::Member` with null parent?